tests/webhook.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/ghodss/yaml"
+
+	"github.com/pingcap/tidb-operator/tests/slack"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/klog"
+)
+
+func (oa *operatorActions) SwitchOperatorWebhook(isWebhookEnabled, isPodWebhookEnable, isStsWebhookEnabled bool, info *OperatorConfig) error {
+	klog.Infof("upgrading tidb-operator with admission webhook %v,pod %v,sts %v", isWebhookEnabled, isPodWebhookEnable, isStsWebhookEnabled)
+
+	//listOptions := metav1.ListOptions{
+	//	LabelSelector: labels.SelectorFromSet(
+	//		label.New().Labels()).String(),
+	//}
+	//pods1, err := oa.kubeCli.CoreV1().Pods(metav1.NamespaceAll).List(listOptions)
+	//if err != nil {
+	//	return err
+	//}
+	switchWebhook := fmt.Sprintf("%s=%v,%s=%v,%s=%v",
+		"admissionWebhook.create", isWebhookEnabled,
+		"admissionWebhook.hooksEnabled.pds", isPodWebhookEnable,
+		"admissionWebhook.hooksEnabled.statefulSets", isStsWebhookEnabled)
+
+	setString := info.OperatorHelmSetString(nil)
+	setString = strings.Join([]string{setString, switchWebhook}, ",")
+
+	cmd := fmt.Sprintf(`helm upgrade %s %s --set-string %s `,
+		info.ReleaseName,
+		oa.operatorChartPath(info.Tag),
+		setString)
```

Comment:
add webhook settings in OperatorCfg and use `UpgradeOperator`?

Reply:
> add webhook settings in OperatorCfg and use `UpgradeOperator`?

That's a bit of hard cause we need to set cabundle when kube-apiserver under 1.13 version which would be complex to embed into `UpgradeOperator`

Reply:
is it possible to configure `caBundle` via `--set-string`? if it is, we can configure `caBundle` in OperatorCfg and update `OperatorConfig.OperatorHelmSetString` to use it if it exists, e.g.

- configure webhook configurations in OperatorConfig
- (if < 1.13) get caBundle from `kube-system.extension-apiserver-authentication`
- (if < 1.13) configure caBundle in OperatorConfig
- UpgradeOperator(OperatorConfig)

Reply:
updated.

Reply:
I set cabundle by `--set-string` at first and encounter a multi-line problem as the cabundle content is multi-line. I used `-f` to avoid this problem.

Reply:
you can fix the multi-line problem by escaping the value of the `--set-string` flag.

Reply:
The current code is fragile because it does not escape the shell arguments. another way is pass command and its arguments to `exec.Command` (don't use `sh`, then no need to escape shell arguments).

Reply:
updated, for now, the switch of each hook and admission server and the cabundle if managed by `OperatorConfig`.